### PR TITLE
Migrate from Webpack to Vite

### PR DESCRIPTION
**Pull request by feihong**
- [x] Remove Webpack dependencies
- [x] Remove all mentions of Webpack
- [x] Remove section about adding configuration for styling (Vite has built-in support for CSS modules)
- [x] Delete all app-specific Makefiles
- [x] Rewrite subsection on Make
- [x] Add new section about bundling to the Better Sandwiches chapter
- [x] Add `index.html` file in every app directory and link to it from root directory's `index.html` file
- [x] Add more detail about where `Index.js` gets generated to 

Needs to be merged at the same time as https://github.com/melange-re/melange-for-react-devs-template/pull/1

### CODE REVIEW

File: index.html
Comment (jchavarri) [inline index.html:1]: Why is this file needed?

File: vite.config.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  plugins: [nodeResolve()],
+  server: {
+    watch: {
+      ignored: ['**/_opam']
+    }
+  },
+  build: {
+    rollupOptions: {
+      input: {
+        better_sandwiches: resolve(__dirname, 'src/better-sandwiches/index.html'),
```
Comment (jchavarri) [inline vite.config.js]: I don't understand why this is only needed for one chapter.

File: docs/intro-to-dune/index.md [hunk 2]
```diff
@@ -150,80 +173,93 @@ src/counter
...
 
-You might be wondering why we're using a `Makefile` instead of adding another
-npm script. In OCaml projects, usage of [the Make build automation
-tool](https://en.wikipedia.org/wiki/Make_(software)) is common, and this
-includes Melange-related projects. So let's take a short detour to explain the
-`Makefile` we just added.
+Now run `make serve` to check that everything works as expected.
 
-## Anatomy of Makefile
+## Why use Make?
```
Comment (jchavarri) [inline docs/intro-to-dune/index.md:205]: (Out of scope) Seeing how much complexity it involves to have both make and npm scripts, I wonder if we could just install the opam switch through some script in npm, e.g. `npm run init-opam` or something like that.

File: src/better-sandwiches/Makefile [hunk 3]
```diff
@@ -1,2 +0,0 @@
-serve:
-	app=better-sandwiches make -C ../.. serve
```
Comment (jchavarri) [inline src/better-sandwiches/Makefile:2]: Why are the submakes not necessary anymore?

**Review by jchavarri — COMMENTED** [review]
There are a few missing references to webpack that I could find:

```
4 results - 2 files

docs/counter/index.md:
  10  1. Run `make watch` to start the Melange compiler in watch mode.
  11: 1. In another terminal window, start the Webpack dev server by running `make
  12  serve`. As a side effect, it will open a browser tab pointed to

docs/styling-with-css/index.md:
   36  Unfortunately, in the terminal where we're running `make serve`, we see this
   37: Webpack compilation error:
   38  

   45  
   46: The problem is that Webpack is serving the app from the build directory at
   47  `_build/default/src/order-confirmation/output/src/order-confirmation`, and the

  245  <b>1.</b> Changing `%%mel.raw` to `%mel.raw` will cause a compile error in
  246: Webpack because the generated JS code changes to
  247  
```

File: index.html
Comment (feihong) [reply index.html:1]: The `make serve` task starts Vite in the root directory, and it expects to serve an `index.html` file in the same directory. I don't think it's necessary to link to the other `index.html` files, but it's nice to provide a home page that links to all of the apps.

File: vite.config.js (hunk 1 above)
Comment (feihong) [reply vite.config.js]: This was left in from where I explain how to make a bundle so you can compare the size with and without `Printf`. But I think it's a bit confusing to leave it in, so I delete it in https://github.com/melange-re/melange-for-react-devs/pull/16/commits/5f3e93ed7834d8d28dd81f7c6e3c0ca57f44cbf7

File: vite.config.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { resolve } from 'path'
+import { defineConfig } from 'vite'
+import { nodeResolve } from '@rollup/plugin-node-resolve'
+
+export default defineConfig({
+  plugins: [nodeResolve()],
```
Comment (feihong) [inline vite.config.js:9]: From https://github.com/melange-re/melange-for-react-devs-template/pull/1/files/afb748d3a47351917943ed43fd0376901a111aff#r1450604037

> I asked in the other PR, but why is nodeResolve needed?

Without it, Vite is unable to import the modules located in, say, `_build/default/src/counter/output/node_modules`. The error you would get is:

```
10:19:45 AM [vite] Internal server error: Failed to resolve import "melange.js/caml_option.js" from "_build/default/src/counter/output/src/counter/Index.js". Does the file exist?
  Plugin: vite:import-analysis
  File: /home/fhsu/work/melange-for-react-devs/_build/default/src/counter/output/src/counter/Index.js:3:31
  1  |  // Generated by Melange
  2  |  
  3  |  import * as Caml_option from "melange.js/caml_option.js";
     |                                ^
  4  |  import * as Melange__Counter from "./Counter.js";
  5  |  import * as ReactDom from "react-dom";
      at formatError (file:///home/fhsu/work/melange-for-react-devs/node_modules/vite/dist/node/chunks/dep-V3BH7oO1.js:63765:46)
      at TransformContext.error (file:///home/fhsu/work/melange-for-react-devs/node_modules/vite/dist/node/chunks/dep-V3BH7oO1.js:63759:19)
      at normalizeUrl (file:///home/fhsu/work/melange-for-react-devs/node_modules/vite/dist/node/chunks/dep-V3BH7oO1.js:62034:33)
      at async file:///home/fhsu/work/melange-for-react-devs/node_modules/vite/dist/node/chunks/dep-V3BH7oO1.js:62188:47
      at async Promise.all (index 0)
      at async TransformContext.transform (file:///home/fhsu/work/melange-for-react-devs/node_modules/vite/dist/node/chunks/dep-V3BH7oO1.js:62109:13)
      at async Object.transform (file:///home/fhsu/work/melange-for-react-devs/node_modules/vite/dist/node/chunks/dep-V3BH7oO1.js:64060:30)
      at async loadAndTransform (file:///home/fhsu/work/melange-for-react-devs/node_modules/vite/dist/node/chunks/dep-V3BH7oO1.js:49741:29)
      at async viteTransformMiddleware (file:///home/fhsu/work/melange-for-react-devs/node_modules/vite/dist/node/chunks/dep-V3BH7oO1.js:59342:32)
10:19:45 AM [vite] Pre-transform error: Failed to resolve import "melange.js/curry.js" from "_build/default/src/counter/output/src/counter/Counter.js". Does the file exist?
```
I guess the default behavior of rollup is to ignore everything in `node_modules`, and this plugin allows you to override that.

File: src/better-sandwiches/Makefile (hunk 3 above)
Comment (feihong) [reply src/better-sandwiches/Makefile:2]: The project-specific Makefile is no longer necessary because of [Vite's multi-page app feature](https://vitejs.dev/guide/build#multi-page-app). They've essentially been replaced by project-specific index.html files.

File: docs/intro-to-dune/index.md (hunk 2 above)
Comment (feihong) [reply docs/intro-to-dune/index.md:205]: I can create a separate PR to migrate from Make to npm scripts after this is merged

**Review by feihong — COMMENTED** [review]
> There are a few missing references to webpack that I could find

Fixed in https://github.com/melange-re/melange-for-react-devs/pull/16/commits/3e3025300104fde3bcf4e0dd4cde6c907552bb4d

File: vite.config.js (hunk 4 above)
Comment (feihong) [reply vite.config.js:9]: @jchavarri Should I add a comment here?

Comment (jchavarri) [reply vite.config.js:9]: Yeah, I think a comment would be helpful. I double checked with @davesnx in case there were other approaches to solve this issue, but apparently this is the way to go (even if it's weird for me to use a Node plugin for an app that is built for the browser...)

**Review by jchavarri — APPROVED** [review]

File: vite.config.js (hunk 4 above)
Comment (jchavarri) [reply vite.config.js:9]: Actually, maybe we should use `browser:true` here? https://github.com/rollup/plugins/tree/master/packages/node-resolve/#browser

Comment (feihong) [reply vite.config.js:9]: @jchavarri Sorry, I read the docs for `browser` option and I still don't quite understand what it does :disappointed:

Comment (feihong) [reply vite.config.js:9]: Added comment in https://github.com/melange-re/melange-for-react-devs/pull/16/commits/bd09d27b7f159197740485600d186e02df83fd01

Comment (jchavarri) [reply vite.config.js:9]: It refers to [this field](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#browser) in `package.json` which will be used by JS tooling like vite to resolve packages that are being bundled for the browser (when a given package has different implementations for Node and browser).
